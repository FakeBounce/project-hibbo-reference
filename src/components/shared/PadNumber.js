import React, { Component } from 'react';
import {
  string,
  func,
  number,
  bool,
  object,
  array,
  oneOfType,
} from 'prop-types';
import { View } from 'react-native';
import EStyleSheet from 'react-native-extended-stylesheet';
import { bind } from 'decko';
import shallowEqual from 'shallowequal';

import erasePic from 'assets/padnumber/erase.png';
import pointPic from 'assets/padnumber/point.png';
import PadNumberItem from './PadNumberItem';

const styles = EStyleSheet.create({
  numberPad: {
    flexWrap: 'wrap',
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'row',
  },
  erase: {
    resizeMode: 'contain',
    width: 30,
    height: 25,
  },
});

class PadNumber extends Component {
  constructor(props) {
    super(props);

    this.numberAfterPoint = 0;
    this.limitPointNumber = false;
    this.amount = this.props.amount || '';
  }

  componentWillReceiveProps = nextProps => {
    if (nextProps.amount !== this.amount) {
      this.amount = nextProps.amount || '';
    }
  };

  shouldComponentUpdate(nextProps, nextState) {
    if (
      shallowEqual(this.state, nextState) &&
      shallowEqual(this.props, nextProps)
    ) {
      return false;
    }

    return true;
  }

  inputDelete(ln) {
    if (ln > 0) {
      if (this.limitPointNumber) {
        if (this.numberAfterPoint <= 0) {
          this.limitPointNumber = false;
          this.numberAfterPoint = 0;
        }
        this.numberAfterPoint--;
      }

      this.amount = this.amount.substr(0, ln - 1);
      if (this.amount === '0') {
        this.amount = '';
      }
      this.props.pressInput(this.amount);
    }
  }

  inputComa(ln) {
    if (this.amount.indexOf(',') === -1) {
      if (ln === 0) {
        this.amount = '0';
      }
      this.amount += ',';
      this.limitPointNumber = true;
      this.numberAfterPoint = -1;
    }
  }

  @bind
  parseInput(input) {
    const ln = this.amount.length;
    const { numberMax, numberAfterComaMax } = this.props;

    if (input === '<') {
      return this.inputDelete(ln);
    }
    if (
      (ln >= numberMax && !this.limitPointNumber && input !== '.') ||
      (this.limitPointNumber && this.numberAfterPoint >= numberAfterComaMax)
    ) {
      return false;
    }

    switch (input) {
      case '0':
        if (ln > 0 || this.props.password) {
          this.amount += '0';
        }
        break;
      case '.':
        this.inputComa(ln);
        break;
      default:
        this.amount += input.toString();
    }

    if (this.amount.length === ln) {
      return false;
    }

    if (this.limitPointNumber) {
      this.numberAfterPoint++;
    }

    return this.props.pressInput(this.amount);
  }

  render() {
    const { style, password } = this.props;
    const numberPad = ['1', '2', '3', '4', '5', '6', '7', '8', '9'];

    return (
      <View style={[styles.numberPad, style]}>
        {numberPad.map(value => (
          <PadNumberItem
            value={value}
            key={value}
            parseInput={this.parseInput}
          />
        ))}
        {!password ? (
          <PadNumberItem
            value="."
            image={pointPic}
            parseInput={this.parseInput}
            removeUnderScore
          />
        ) : (
          <PadNumberItem
            value=""
            parseInput={() => {}}
            removeUnderScore
            visible={false}
          />
        )}
        <PadNumberItem
          value="0"
          parseInput={this.parseInput}
          removeUnderScore
        />
        <PadNumberItem
          value="<"
          image={erasePic}
          imgStyle={styles.erase}
          parseInput={this.parseInput}
          removeUnderScore
        />
      </View>
    );
  }
}

PadNumber.defaultProps = {
  amount: '',
  style: 0,
  password: false,
  numberMax: 6,
  numberAfterComaMax: 2,
};

PadNumber.propTypes = {
  amount: string,
  pressInput: func.isRequired,
  style: oneOfType([object, number, array]),
  password: bool,
  numberMax: number,
  numberAfterComaMax: number,
};

export default PadNumber;
