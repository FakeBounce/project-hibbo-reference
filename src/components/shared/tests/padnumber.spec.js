import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import PadNumber from 'shared/PadNumber';

describe('PadNumber', () => {
  it('should render the component', () => {
    const pressInput = jest.fn();

    const padNumber = renderer
      .create(<PadNumber pressInput={pressInput} />)
      .toJSON();

    expect(padNumber).toMatchSnapshot();
  });

  it('should click the component', () => {
    const pressInput = jest.fn();

    const tree = shallow(<PadNumber pressInput={pressInput} />);
    const res = '123456';

    for (let i = 0; i < tree.find('PadNumberItem').length; ++i) {
      tree
        .find('PadNumberItem')
        .at(i)
        .shallow()
        .find('TouchableRipple')
        .first()
        .shallow()
        .simulate('press');
      expect(pressInput).toHaveBeenCalled();
      if (i < 9) {
        expect(pressInput).toHaveBeenLastCalledWith(res.substr(0, i + 1));
      } else if (i === 10) {
        expect(pressInput).toHaveBeenLastCalledWith(`${res},0`);
      } else {
        expect(pressInput).toHaveBeenLastCalledWith(`${res},`);
      }
    }

    expect(pressInput).toHaveBeenCalledTimes(9);
  });
});
