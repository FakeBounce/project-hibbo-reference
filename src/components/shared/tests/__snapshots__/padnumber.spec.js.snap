// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PadNumber should render the component 1`] = `
<View
  style={
    Array [
      undefined,
      0,
    ]
  }
>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={Object {}}
    testID={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={undefined}
      >
        1
      </Text>
      <Image
        source={1}
        style={undefined}
      />
    </View>
    <View
      style={
        Array [
          undefined,
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "backgroundColor": "#303030",
            "borderRadius": 70,
            "height": 70,
            "opacity": 0,
            "transform": Array [
              Object {
                "scale": 0.4,
              },
            ],
            "width": 70,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={Object {}}
    testID={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={undefined}
      >
        2
      </Text>
      <Image
        source={1}
        style={undefined}
      />
    </View>
    <View
      style={
        Array [
          undefined,
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "backgroundColor": "#303030",
            "borderRadius": 70,
            "height": 70,
            "opacity": 0,
            "transform": Array [
              Object {
                "scale": 0.4,
              },
            ],
            "width": 70,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={Object {}}
    testID={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={undefined}
      >
        3
      </Text>
      <Image
        source={1}
        style={undefined}
      />
    </View>
    <View
      style={
        Array [
          undefined,
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "backgroundColor": "#303030",
            "borderRadius": 70,
            "height": 70,
            "opacity": 0,
            "transform": Array [
              Object {
                "scale": 0.4,
              },
            ],
            "width": 70,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={Object {}}
    testID={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={undefined}
      >
        4
      </Text>
      <Image
        source={1}
        style={undefined}
      />
    </View>
    <View
      style={
        Array [
          undefined,
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "backgroundColor": "#303030",
            "borderRadius": 70,
            "height": 70,
            "opacity": 0,
            "transform": Array [
              Object {
                "scale": 0.4,
              },
            ],
            "width": 70,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={Object {}}
    testID={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={undefined}
      >
        5
      </Text>
      <Image
        source={1}
        style={undefined}
      />
    </View>
    <View
      style={
        Array [
          undefined,
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "backgroundColor": "#303030",
            "borderRadius": 70,
            "height": 70,
            "opacity": 0,
            "transform": Array [
              Object {
                "scale": 0.4,
              },
            ],
            "width": 70,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={Object {}}
    testID={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={undefined}
      >
        6
      </Text>
      <Image
        source={1}
        style={undefined}
      />
    </View>
    <View
      style={
        Array [
          undefined,
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "backgroundColor": "#303030",
            "borderRadius": 70,
            "height": 70,
            "opacity": 0,
            "transform": Array [
              Object {
                "scale": 0.4,
              },
            ],
            "width": 70,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={Object {}}
    testID={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={undefined}
      >
        7
      </Text>
      <Image
        source={1}
        style={undefined}
      />
    </View>
    <View
      style={
        Array [
          undefined,
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "backgroundColor": "#303030",
            "borderRadius": 70,
            "height": 70,
            "opacity": 0,
            "transform": Array [
              Object {
                "scale": 0.4,
              },
            ],
            "width": 70,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={Object {}}
    testID={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={undefined}
      >
        8
      </Text>
      <Image
        source={1}
        style={undefined}
      />
    </View>
    <View
      style={
        Array [
          undefined,
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "backgroundColor": "#303030",
            "borderRadius": 70,
            "height": 70,
            "opacity": 0,
            "transform": Array [
              Object {
                "scale": 0.4,
              },
            ],
            "width": 70,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={Object {}}
    testID={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={undefined}
      >
        9
      </Text>
      <Image
        source={1}
        style={undefined}
      />
    </View>
    <View
      style={
        Array [
          undefined,
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "backgroundColor": "#303030",
            "borderRadius": 70,
            "height": 70,
            "opacity": 0,
            "transform": Array [
              Object {
                "scale": 0.4,
              },
            ],
            "width": 70,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={Object {}}
    testID={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Image
        source={1}
        style={Object {}}
      />
    </View>
    <View
      style={
        Array [
          undefined,
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "backgroundColor": "#303030",
            "borderRadius": 70,
            "height": 70,
            "opacity": 0,
            "transform": Array [
              Object {
                "scale": 0.4,
              },
            ],
            "width": 70,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={Object {}}
    testID={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={undefined}
      >
        0
      </Text>
    </View>
    <View
      style={
        Array [
          undefined,
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "backgroundColor": "#303030",
            "borderRadius": 70,
            "height": 70,
            "opacity": 0,
            "transform": Array [
              Object {
                "scale": 0.4,
              },
            ],
            "width": 70,
          }
        }
      />
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={Object {}}
    testID={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Image
        source={1}
        style={Object {}}
      />
    </View>
    <View
      style={
        Array [
          undefined,
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "backgroundColor": "#303030",
            "borderRadius": 70,
            "height": 70,
            "opacity": 0,
            "transform": Array [
              Object {
                "scale": 0.4,
              },
            ],
            "width": 70,
          }
        }
      />
    </View>
  </View>
</View>
`;
